components/validator/issuer.go

```diff
@@ -0,0 +1,272 @@
+package validator
+
+import (
+	"context"
+	"time"
+
+	"github.com/iotaledger/hive.go/ierrors"
+	iotago "github.com/iotaledger/iota.go/v4"
+	"github.com/iotaledger/iota.go/v4/builder"
+)
+
+var ErrBlockTooRecent = ierrors.New("block is too recent compared to latest commitment")
+
+type TaskType uint8
+
+const (
+	CandidateTask TaskType = iota
+	CommitteeTask
+)
+
+func candidateAction(ctx context.Context) {
+	now := time.Now()
+	currentAPI := deps.NodeBridge.APIProvider().APIForTime(now)
+	currentSlot := currentAPI.TimeProvider().SlotFromTime(now)
+
+	isCandidate, err := deps.NodeBridge.ReadIsCandidate(ctx, validatorAccount.ID(), currentSlot)
+	if err != nil {
+		Component.LogWarnf("error while checking if account is already a candidate: %s", err.Error())
+		// If there is an error, then retry registering as a candidate.
+		executor.ExecuteAt(CandidateTask, func() { candidateAction(ctx) }, now.Add(ParamsValidator.CandidacyRetryInterval))
+
+		return
+	}
+
+	if isCandidate {
+		Component.LogDebugf("not issuing candidacy announcement block as account %s is already registered as candidate in epoch %d", validatorAccount.ID(), currentAPI.TimeProvider().EpochFromSlot(currentSlot))
+		// If an account is a registered candidate, then try to register as a candidate in the next epoch.
+		executor.ExecuteAt(CandidateTask, func() { candidateAction(ctx) }, currentAPI.TimeProvider().SlotStartTime(currentAPI.TimeProvider().EpochStart(currentAPI.TimeProvider().EpochFromSlot(currentSlot)+1)))
+
+		return
+	}
+
+	if epochNearingThresholdSlot := currentAPI.TimeProvider().EpochEnd(currentAPI.TimeProvider().EpochFromSlot(currentSlot)) - currentAPI.ProtocolParameters().EpochNearingThreshold(); currentSlot > epochNearingThresholdSlot {
```

Comment:
Is `>` correct here, or should it be `>=`?

Reply:
This is correct, because the last possible slot to announce candidacy is EpochEnd - EpochNearingThreshold, not the one before it. After this slot is finalized, then committee is rotated taking into account all candidates registered until  (and including) that slot.